### ISSUE TITLE
In nested Describes, executing loops over a global iterable(map, slice) before It block, does not execute any test

### ISSUE THREAD

**Issue by puppet-py** [onsi/ginkgo#693]
From a ginkgo novice:

Here is my intent: In the package there are 2 global variables of type map & slice resp. 
Accessing the above two global variable, I want to iterate same testcase(It) over a slice which is nested inside a map(nested for loop), thereby generating multiple tests. Each entry inside is a map need to be validated against each entry inside the slice

My test code structure is: 
 Outer Describe { var declarations;  BeforeEach; Inner Describe {Context(accesses Outer Describe's declared vars); nested for loop; It Block inside the Inner for loop}

Observation: No single testcase is getting executed. It returns "Ran 0 of 0 Specs". If you could point me where I am going wrong in my approach , really appreciate the help

Auxiliary doubt: As I intend to run the same It block inside the for loop, does that mean for "n" iterations, ginkgo will treats as "n" specs ?

````
func TestFilterTest(t *testing.T) {
	RegisterFailHandler(Fail)
	By("Setting up the global state of test environment")
	mydatafuncs.BuildTestEnv() //This sets up the global variables inside mydatafuncs.
	RunSpecs(t, " Drone Functionality Test Suite")
}
var _ = BeforeSuite(func() {

	By("Call API for Drone data")
	err := mydatafuncs.GenerateDd()
	Expect(err).NotTo(HaveOccurred())
})
var _ = AfterSuite(func() {
	By("Cleanup: Delete the copied and generated db and json files")
	err := mydatafuncs.DeleteFiles("ds.json", "ts.json", "dt.db")
	Expect(err).NotTo(HaveOccurred())
})

var _ = Describe("Drone datafill Functionality", func() {

	var (
		dataset = mydatafuncs.TestConfig.Dataset
		version = mydatafuncs.TestConfig.DatasetVersion
	)

	BeforeEach(func() {
		By("Deleting the cross.db before running tests" +
			"for each dataset")
		err := mydatafuncs.DeleteFiles("cross.db")
		Expect(err).NotTo(HaveOccurred())
	})
	Describe("Let see what happens ", func() {
		Context(fmt.Sprintf("Apply gauss on tablesets"), func() {
			for gauset, fills := range mydatafuncs.GausOfTs {
				ts := gauset; fltrs := fill //Since we are inside a closure, else only last value will be passed
				for _, filter := range fltrs {
					flt := filter
					It("There is no diffs between two tables", func() {
						result, _ := mydatafuncs.CompareTablesFromDFs(dataset, version,
							ts, flt)
						Expect(result).Should(BeTrue())
					})
				}
			}
		})
	})
})
````

**Comment by onsi** [comment]
hey @puppet-py sorry to say you’ve been bitten by a subtlety in the ginkgo lifecycle (which is sorely needing documentation).  i’ll reply in a bit more detail soon but for now:

the issue is that the `mydatafuncs.BuildTestEnv()` is being called in `TestFilterTest` which is invoked by the go test runner only _after_ the contents of the file have been loaded and executed.

but it is that step of loading and executing the file that calls the `Describe`s and `Context`s that build the testing tree.  so tl;dr your data funcs test environment hasn’t been set up yet.

what you’re trying to accomplish is supported and a common pattern. to fix this you can invoke `mydatafuncs.BuildTestEnv()` in the top-level `Describe` just before the `var` block.  that assumes this is the only describe - if you have others then we’ll need to take a different approach.  let me know if that works for you or gives you trouble!

**Comment by puppet-py** [comment]
As you suggested, I just tried moving `mydatafunc.BuildTestEnv()` out of `TestFilterTest` as below >> (rest of the code block is as)
```
mydatafunc.BuildTestEnv()
var _ = Describe("Drone datafill Functionality", func() {
```

It fails with these messages >>
```
Failed to compile filtertests
expected declaration, found filtertests
```

**Comment by onsi** [comment]
Right, so sadly that won’t work because Go insists on not letting you run bare functions at the top level.  Your function has to return something and you have to assign that something - which is why Ginkgo does this deal:

`var _ = Describe(...`

for the top-level Describe.  That discarded assignment appeases go’s rules about no unassigned top-level function invocations.

so...

if you want what you’re doing to work you have to have `mydatafunc.BuildTestEnv()` return something that you discard (eg `nil`) or you can put `mydatafunc.BuildTestEnv()` in the top level `Describe`:

```
var _ = Describe("Drone datafill Functionality", func() {
        mydatafunc.BuildTestEnv()
        var (
		dataset = mydatafuncs.TestConfig.Dataset
		version = mydatafuncs.TestConfig.DatasetVersion
	)
       ....
```

(sorry if the indentation is off - i’m on my phone)

**Comment by puppet-py** [comment]
@onsi Thanks for being so responsive !!
Well, that was bit silly of me that at the top level I invoked the function `mydatafunc.BuildTestEnv()` without a dummy assignment(that function was earlier doing naked return)

Anyways, I have now, tried this >> (`mydatafunc.BuildTestEnv` is now returning <nil>)
```
var _ = mydatafunc.BuildTestEnv()
var _ = Describe("Drone datafill Functionality", func() {
```
Then I getting this error(very ambiguous) right after the `mydatafunc.BuildTestEnv` is successfully executed >>
```
2020/06/16 21:52:30 Read the Config file
2020/06/16 21:52:30 Check the config file params(raises panic)
2020/06/16 21:52:30 Copy ds.json & ts.json into current test dir(raises panic)
2020/06/16 21:52:30 Generate Filters of each Table from ds.json
flag provided but not defined: -test.timeout
Usage of /Users/puppet-py/workspace/go/src/filtertests/filtertests.test:
  -alsologtostderr
    	log to standard error as well as files (default true)
  -block-profile string
    	filename prefix for block profiles
  -block-profile-rate int
    	rate for runtime.SetBlockProfileRate (default 200)
  -cpu-profile string
<< truncated >>
```
I am getting the same error if I call the `mydatafuncs.BuildTestEnv` inside the Outer Describe as you  provided in your earlier comments.

Sincerely we need better documentation on understanding the flow

**Comment by onsi** [comment]
> Sincerely we need better documentation on understanding the flow

yeah - I hope to work on it soon.

I think you may be hitting this issue here:
https://github.com/onsi/ginkgo/issues/602

Does `mydatafunc.BuildTestEnv()` call `flag.Parse()`?

Also - is the code you're working on published somewhere so I can poke around?

**Comment by puppet-py** [comment]
It does not call `flag.Parse() `nor there is any `init` function. Did not find any relevance other than the error wrt https://github.com/onsi/ginkgo/issues/602

I will try to come up with a dummy code to be very similar to what I have, such that I share it for public review. Will work on it soon

**Comment by onsi** [comment]
ok thanks

**Comment by puppet-py** [comment]
@onsi 
Keeping the exact code workflow as with my actual test, I created a dummy test. The ONLY diff between the two test files are their respective imported functions. 
Observation: the "for-loop" and "global variable" access works fine in my dummy-tests however the actual test continues to give me `flag provided but not defined: -test.timeout` error. I recreated the file with a different name thinking if it has any cache-related issue(could not think any other logic) but I still hit the same error

I have sent you an invite to my private repo for you to have access to the codebase:

https://github.com/puppet-py/ginkgoTest

**Comment by onsi** [comment]
I pulled the repo you pointed me at and did see the dummy tests pass.  I can't run the real tests as I don't have access to the actual `filtertest` package.  Poking around v23 context briefly it does look like it pulls in some logging libraries that seem to take configuration from the command line and import the `flag` package.  while I don't see an explicit call to `flag.Parse` my hypothesis is that it's being called somewhere by something during the invocation of `GetContext()`.

Go's flag parsing seems to be particularly brittle around the ordering of things.  If you can replace v23 context (or hack at it to confirm/remove the invocation of `flag.Parse` that would likely get you past this issue.

If you can't - there's one more thing to try.  The test suite structure needs to be constructed prior to invoking `RunSpecs`.  Typically the test suite is constructed as the binary loads and invokes all the top-level things.  Unfortunately the rough order looks to be:

- invoke all the top-level things, thereby constructing the test suite
- run the `func init()`s which is where go test now invokes `flag.Parse()` (this means any prior invocation of `flag.Parse` in the top-level step will fail)
- go test then runs the `TestFilterTest` which...
- ...calls `RunSpecs(...)`

So I'm pretty sure you could do something like this:

```
func TestFilterTest(t *testing.T) {
	RegisterFailHandler(Fail)
	mydatafuncs.BuildTestEnv() //This sets up the global variables inside mydatafuncs.
        BuildTestSuite() //see below, this constructs the test suite
	RunSpecs(t, " Drone Functionality Test Suite")
}

func BuildTestSuite() {
	BeforeSuite(func() {
		By("Call API for Drone data")
		err := mydatafuncs.GenerateDd()
		Expect(err).NotTo(HaveOccurred())
	})
	AfterSuite(func() {
		By("Cleanup: Delete the copied and generated db and json files")
		err := mydatafuncs.DeleteFiles("ds.json", "ts.json", "dt.db")
		Expect(err).NotTo(HaveOccurred())
	})

	Describe("Drone datafill Functionality", func() {
		var (
			dataset = mydatafuncs.TestConfig.Dataset
			version = mydatafuncs.TestConfig.DatasetVersion
		)

		BeforeEach(func() {
			By("Deleting the cross.db before running tests" +
				"for each dataset")
			err := mydatafuncs.DeleteFiles("cross.db")
			Expect(err).NotTo(HaveOccurred())
		})
		Describe("Let see what happens ", func() {
			Context(fmt.Sprintf("Apply gauss on tablesets"), func() {
				for gauset, fills := range mydatafuncs.GausOfTs {
					ts := gauset; fltrs := fill //Since we are inside a closure, else only last value will be passed
					for _, filter := range fltrs {
						flt := filter
						It("There is no diffs between two tables", func() {
							result, _ := mydatafuncs.CompareTablesFromDFs(dataset, version,
								ts, flt)
							Expect(result).Should(BeTrue())
						})
					}
				}
			})
		})
	})
}
```

Now the order is:

- invoke all the top-level things (this is a no-op, it just defines a few functions)
- run the `func init()`s which is where go test now invokes `flag.Parse()` (nothing has called `flag.Parse()` yet so all should be well)
- go test then runs the `TestFilterTest` which...
- calls `BuildTestSuite()` which constructs the testing tree and then...
- ...calls `RunSpecs(...)`

The reason this isn't a common pattern in Ginkgo is because we try to support having multiple test files within a given package.  With this approach you'd have to build functions like this in each file and ensure you invoke them in the `TestFilterTest` - the default behavior makes things a bit more transparent/seamless.  It works well for the 90% case but does make the 10% case where the lifecycle becomes important harder to reason about.

I'll think through if there's a better design that could help folks avoid gotchas like this.

Um... last disclosure... I, er, haven't actually _tried_ this yet 😅 - could you give it a shot and confirm that it does in fact work?

**Comment by onsi** [comment]
Just tried it and verified that the approach I'm proposing works.  Let me know if it does/doesn't @puppet-py 

This thread has also given me some ideas for how to prevent this particular class of gotchas.  I'll explore them and see if any merit addition to Ginkgo.

**Comment by onsi** [comment]
Sorry - didn't mean to close the issue but I just pushed a change to master that defers calling the  bodies of the  top-level `Describes`until `RunSpecs` is called.  This will allow your original code to work and should make the lifecycle a bit easier for folks to reason about.

Still need to document it all, though!

**Comment by puppet-py** [comment]
@onsi The order which you have outlined above worked like a charm ! Thanks much again.

**Comment by puppet-py** [comment]
> to master that defers calling the bodies of the top-level `Describes`until `RunSpecs` is called. This will allow your original code to work and should make the lifecycle a bit easier for folks to reason about

Looking forward to it. Will the commit be available on package update or should I have to wait for a release date ?

**Comment by onsi** [comment]
Haven't cut a new tag yet but it's on master.  Will probably cut a tag
sometime next week.

Onsi

**Comment by puppet-py** [comment]
@onsi 
I don't see the new tag comprising the fix/changes pertaining to this issue. If you have not dont yet, I sincerely request if you can get the following done as early the time permits you :
- tag
- documentation

**Comment by onsi** [comment]
Ah - sorry, I was out on vacation last week.  I'll try to ship a tag in the
next few days (most likely Thursday).

Onsi

**Comment by onsi** [comment]
Released v1.14.0 which includes this.  Also, added documentation here:
http://onsi.github.io/ginkgo/#understanding-ginkgos-lifecycle
